packages/builder/src/components/backend/Datasources/CreateEditRelationship.svelte

```diff
@@ -116,93 +90,90 @@
-    if (invalidThroughTable()) {
-      errObj.throughTable =
-        "Ensure non-key columns are nullable or auto-generated"
-    }
-    if (!isMany && !fromForeign) {
-      errObj.fromForeign = foreignKeyNotSet
-    }
-    if (!fromColumn) {
-      errObj.fromColumn = colNotSet
-    }
-    if (!toColumn) {
-      errObj.toColumn = colNotSet
-    }
-    if (!isMany && !fromPrimary) {
-      errObj.fromPrimary = "Please pick the primary key"
-    }
-    if (isMany && relationshipExists()) {
-      errObj.fromTable = relationshipAlreadyExists
-      errObj.toTable = relationshipAlreadyExists
-    }
-
-    // currently don't support relationships back onto the table itself, needs to relate out
-    const tableError = "From/to/through tables must be different"
-    if (fromTable && (fromTable === toTable || fromTable === throughTable)) {
-      errObj.fromTable = tableError
-    }
-    if (toTable && (toTable === fromTable || toTable === throughTable)) {
-      errObj.toTable = tableError
-    }
+  function relationshipExists() {
     if (
-      throughTable &&
-      (throughTable === fromTable || throughTable === toTable)
+      originalFromTable &&
+      originalToTable &&
+      originalFromTable === fromTable &&
+      originalToTable === toTable
     ) {
-      errObj.throughTable = tableError
-    }
-    const colError = "Column name cannot be an existing column"
-    if (isColumnNameBeingUsed(toTable, fromColumn, originalFromColumnName)) {
-      errObj.fromColumn = colError
-    }
-    if (isColumnNameBeingUsed(fromTable, toColumn, originalToColumnName)) {
-      errObj.toColumn = colError
+      return false
     }
+    let fromThroughLinks = Object.values(
+      datasource.entities[fromTable.name].schema
+    ).filter(value => value.through)
+    let toThroughLinks = Object.values(
+      datasource.entities[toTable.name].schema
+    ).filter(value => value.through)
 
-    let fromType, toType
-    if (fromPrimary && fromForeign) {
-      fromType = fromTable?.schema[fromPrimary]?.type
-      toType = toTable?.schema[fromForeign]?.type
-    }
-    if (fromType && toType && fromType !== toType) {
-      errObj.fromForeign =
-        "Column type of the foreign key must match the primary key"
-    }
+    const matchAgainstUserInput = (fromTableId, toTableId) =>
+      (fromTableId === fromId && toTableId === toId) ||
+      (fromTableId === toId && toTableId === fromId)
 
-    errors = errObj
-    return getErrorCount(errors) === 0
+    return !!fromThroughLinks.find(from =>
+      toThroughLinks.find(
+        to =>
+          from.through === to.through &&
+          matchAgainstUserInput(from.tableId, to.tableId)
+      )
+    )
   }
 
-  function isColumnNameBeingUsed(table, columnName, originalName) {
-    if (!table || !columnName || columnName === originalName) {
-      return false
+  function getErrorCount(errors) {
+    return Object.entries(errors)
+      .filter(entry => !!entry[1])
+      .map(entry => entry[0]).length
```

Comment:
Why do we need the **map** if we're just getting the length?

Reply:
Unsure - will remove it, just moved that code around!